src/event-log/event-log.ts

```diff
@@ -0,0 +1,32 @@
+export type Event = {
+  watermark: string,
+  messageCid: string
+};
+
+export interface EventLog {
+ /**
+  * opens a connection to the underlying store
+  */
+  open(): Promise<void>;
+
+  /**
+   * closes the connection to the underlying store
+   */
+  close(): Promise<void>;
+
+  /**
+   * adds an event to a tenant's event log
+   * @param tenant - the tenant's DID
+   * @param messageCid - the CID of the message
+   */
+  append(tenant: string, messageCid: string): Promise<string>
```

Comment:
i wonder if we should have this take an actual `BaseMessage` instead of a `messageCid` to ensure that we're actually getting the entire message and not just some random CID (similar to how `MessageStore` requires the entire `BaseMessage` for `put`)

also, how does this handle stuff like `deleteAllOlderMessagesButKeepInitialWrite` where we remove any overridden `RecordsWrite` (except for the initial message and the newest one)?  all the intervening messages will no longer exist in the `MessageStore`, so having entries for them in the `EventLog` seems a bit odd (and possibly problematic if a new `RecordsWrite` is received in between an `append` and `getEventsAfter` as the other `Dwn` would have a `messageCid` that no longer exists in the `MessageStore` since it would've been deleted by that new `RecordsWrite`)

Reply:
@dcrousso adding a visualization here to illustrate what you're talking about.

<img width="1505" alt="image" src="https://user-images.githubusercontent.com/4887440/229679820-964474e4-2434-4d80-984d-98a39aa656af.png">

then thinking out loud now...  🤔 

any time an incoming `RecordsWrite` or `RecordsDelete` targets a pre-existing `recordId`, all intermediate `RecordsWrites` _except_ for the _initial_ `RecordsWrite` _and_ the _latest_ message (which could be a `RecordsWrite` _or_ a `RecordsDelete`) are dropped. 

the response of any `EventsGet` _after_ `t=2` will end up including events with `messageCid`s that no longer exist. 

i suppose the path of least resistance to accommodate this is to delete the events associated to those dropped messages. Instinctively (with no logical reason to back it), deleting events from a log that's supposed to keep track of _all_ persisted messages feels off.. but the DWN design seems to be alright with that, so, it sounds like i need to implement a `deleteEvent` or `deleteEvents`

Reply:
tried my best to ignore this problem out of existence but it didn't work

Reply:
would like to do something like this:

```js
async function deleteEventsByCid(tenant, cids) {
  const cidSet = new Set(cids);
  const tenantEventLog = this.db.partition(tenant);
  const ops = [];

  for await (const [key, value] of tenantEventLog.iterator()) {
    if (cidSet.has(value)) {
      ops.push({ type: 'del', key });
    }
  }

  await this.tenantEventLog.batch(ops);
}
```

the inefficiency here is that we have to iterate through a tenant's _entire_ event log to find all CIDs. this inefficiency is compounded given the way that [deleteAllOlderMessagesButKeepInitialWrite](https://github.com/TBD54566975/dwn-sdk-js/blob/main/src/interfaces/records/records-interface.ts#L15-L42) is written and more generally how [RecordsWriteHandler](https://github.com/TBD54566975/dwn-sdk-js/blob/main/src/interfaces/records/handlers/records-write.ts#L24-L103) and [RecordsDeleteHandler](https://github.com/TBD54566975/dwn-sdk-js/blob/main/src/interfaces/records/handlers/records-delete.ts#L17-L77) are implemented.

imo, there's quite a bit in those 3 areas of code that can be simplified and optimized. probably best to leave this simplify & optimize refactor for a separate PR, so i'll implement `deleteEventsByCid` as shown above and slot it into the `deleteAllOlderMessagesButKeepInitialWrite` loop. 

lmk if anyone else has other ideas

Reply:
for now could/should we just have it be `delete(tenant: string, messageCid: string): Promise<void>`?  i.e. only do one message at a time?

(i mentioned it offline, but ill put it here for future reference just in case) if we're OK with the storage cost we could also have it include a reverse lookup (i.e. `messageCid` -> `watermark`) to make the deletion much faster (i.e. no need to `iterator` over everything, just one `get` and `delete`)

i definitely appreciate the desire to use `batch`, but unless we include the `watermark` in the `Message` itself (and require that the whole `Message` be provided for `delete`) i dont think we'll be able to easily avoid some sort of lookup somewhere to get the `watermark` in order to delete it :/

Reply:
oh hmm here's an idea i just had :D

what if instead of doing this when new messages come in, we lazily do this inside `getEventsAfter`?

i.e. instead of `delete` whenever we remove something from the `MessageStore`, instead we wait until someone calls `getEventsAfter` and do a cleanup whenever we notice that an override/replacement/delete/etc. happened?  we already have to `iterator` inside `getEventsAfter` so why not leverage that to get the `watermark` key while we're at it?  and once we've finished the `iterator` we can then do a cleanup of the resulting `events` to find things that we need to delete and submit one final `batch` to remove them all?

Reply:
> for now could/should we just have it be `delete(tenant: string, messageCid: string): Promise<void>`? i.e. only do one message at a time?
> 
> (i mentioned it offline, but ill put it here for future reference just in case) if we're OK with the storage cost we could also have it include a reverse lookup (i.e. `messageCid` -> `watermark`) to make the deletion much faster (i.e. no need to `iterator` over everything, just one `get` and `delete`)
> 
> i definitely appreciate the desire to use `batch`, but unless we include the `watermark` in the `Message` itself (and require that the whole `Message` be provided for `delete`) i dont think we'll be able to easily avoid some sort of lookup somewhere to get the `watermark` in order to delete it :/

I think i can still keep `deleteEventsByCid` as it is and just pass it a single cid array for now.

Reply:
@dcrousso [this commit](https://github.com/TBD54566975/dwn-sdk-js/pull/290/commits/74a91abd8c17a2c60395f87ccd22b5ac71432071) remedies the original point you made

Reply:
For me, I think it is okay for sync to not able to get a message of a given CID, and we can treat the message being obsolete as the (only?) expected condition of a message fetch to return nothing. There is probably no easy way to prevent this condition to occur: even if we delete the associated events at the time a message is dropped, because a message could be dropped right after `getEvents()` gets called causing immediate message fetch by CID to "fail".

I see deleting the events of deleted messages as an optimization, one dumb place to do it is on demand when fetching message by CID fails. Though like your initial thoughts, I have mixed feelings about deleting "events" from log, but coming around to it because keeping it serves no practical purposes and is just a long-term perf and storage drag.

Reply:
i disagree with the notion that deleting events that aren't present at time of delete or overwrite is an "optimization".  

we cannot assume that a CID not being present _always_ implies that the CID has been deleted. this means that sync _must always_ fetch _all_ data associated to messages. a situation where that's not feasible is 4gb blu-ray chonkers.

Reply:
cc: @csuwildcat

Reply:
Hm... maybe I misread the design, I thought the event log keeps track of _message_ CIDs? If so, isn't data CIDs mostly irrelevant to the discussion? When DWN returns a message CID from even log, but fails to give you the actual message when you try to fetch it by message CID, it seems fine to assume that it is being deleted no? If we don't love the implicitness, we could return an explicit "return code" that says "deleted". But regardless of the explicitness/implicitness, data CID(s) seems to be a layer of abstraction away, not seeing data associated with the message needing to always be downloaded by a DWN. Am I smocking crack...?!

Reply:
naw you're right @thehenrytsai i'm the one smocking crack. it's message CID based. had something else on my mind and was delirious by the time i wrote this

Reply:
see https://github.com/TBD54566975/dwn-sdk-js/pull/290#issuecomment-1500738602